rkt: Implement distribution points

sgotti opened the pull request:
This is the implementation of the distribution concept proposed in #2953.
## 

This patch implements the distribution point concept with the three initial
kind of distributions already used by rkt (appc, aci-archive, docker).

It just makes the smallest required changes to the fetchers logic to make them
work and pass the functional tests.

This is the base for future fetchers (a future patch will refactor the
fetcher logic to better fit the distribution concept) and store refactors.

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+	// Returns a copy of the distribution URI
+	URI() *url.URL
+	// Returns the distribution type
+	Type() DistType
```
The distribution type is encoded in two places, as `distribution.Docker`, and `DistTypeDocker`. I am not convinced we need this type information in two places.

sgotti replied:
I was thinking to just use type assertions/switches and remove DistType and Type() from the Distribution interface. Or do you have other suggestions?

s-urbaniak replied:
I was hoping that we could leverage polymorphic behavior by using interfaces and avoid type assertions/switches as much as possible. I am hesitant to use DistType switches in upper architectual layers (although I am aware this is hard).

sgotti replied:
> I am hesitant to use DistType switches in upper architectual layers (although I am aware this is hard).

I'd also like to avoid this if possible. Some thoughts:

* Upper code needs a way to know the distribution type. e.g. for calling the right distribution handler. Possible solutions:

 * Keep the `Type() DistType` func from the Distribution interface as now.
 * Use an helper function `DistType(d Distribution) (DistType, error)` (removing the `Type() DistType` func from the Distribution interface)

* To avoid type assertions I can remove all the additional (not covered by the Distribution interface) methods from the distributions (`Appc.App() *discovery.App` for Appc, `ACIArchive.ArchiveURL() *url.URL` for ACIArchive, `Docker.DockerString()` `Docker.SimpleDockerString() string` for Docker) and implement generic helpers methods like 
```
AppDiscoveryString(d Distribution) (*discovery.App,error) // Will work only for Appc Distribution
TransportURL(d Distribution) (*url.URL, error) // Will work only for ACIArchive and perhaps we can use the same function for future distributions like oci-image-layout
DockerString(d Distribution) (string,error) // Will work only for Docker Distribution
SimpleDockerString(d Distribution) (string,error) // Will work only for Docker Distribution
```
Since they will have a Distribution interface as input (to avoid type assertions) they have to do validation since they'll work only for a specific distribution type (so they will also return an error) (and quite probably they'll do an internal type assertion).

This will avoid the need for the user to know the additional provided methods of a distribution type but just use some helper methods.

Thoughts? Better Ideas?

s-urbaniak replied:
This is nice improvement :+1: One question regarding the "distribution handler": Couldn't this also be an interface type included in the returned struct from a potential `Parse(...) ...` method? Just fighting for more generalization ;-)

sgotti replied:
>  One question regarding the "distribution handler": Couldn't this also be an interface type included in the returned struct from a potential Parse(...) ... method? Just fighting for more generalization ;-)

The distribution handlers, like the user friendly string and the use of distribution URIs as references in #3071 are on top of this. Every user could implement their own logic. Given this, I'd like that this could be shared be more projects since it could probably be of help also for k8s (@euank @yifan-gu) for handling multiple diatribution types (sometimes called, IMHO not correctly, image types).

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+// A Distribution represnt the way to retrieve an image starting from an input
+// string.
+// It's identified by an URI with a specific schema
+type Distribution interface {
```
This interface does not define any behavior, just data.

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+	DistScheme = "cimd"
+
+	// Distribution types
+	DistTypeAppc       DistType = "appc"
```
Rather than specifying strings here, why not go with concrete distribution instances, and let the `Distribution` interface implement only behavior, i.e `Compare(Distribution) bool`?

Compare https://github.com/opencontainers/image-spec/blob/master/schema/schema.go#L21.

sgotti replied:
I was thinking that forcing a Distribution to implement `ComparableString() string` will add two things in one:

* Have a ready to use and reproducible distribution serialization string.
* avoid them implementing a Compare method and use the common Compare function.

But these can also be done removing the `ComparableString() string` method from the interface and add a `Compare(Distribution) bool` method. For distribution serialization just URI().String() could be used (so an user isn't tempted to just compare two distribution string but use the `Compare` method.
It'll probably be cleaner. Will experiment with this.

s-urbaniak replied:
sounds good, thanks a lot for the patience :+1: :-)

sgotti commented:
Updated the PR.

* The `Distribution` interface has only two methods (`URI()` and `Compare(Distribution)`)
* Now every distribution will register itself. This avoids a type switch inside `New(...)` and `Parse(...)` and makes easy to add new distributions.
* Removed all the custom distributions functions and added generic helper functions. This avoid doing type assertions at the cost of an increased number of error checks.

@s-urbaniak I tried to simplify as much as possible. An extreme simplification will be to just remove the `Distribution` interface and declare `type Distribution *uri.URI`. This will require, since comparison is a primary property of a distribution, that the URI string will be provided ready to be compared. But I'd prefer the current way since an interface forces the right behavior instead of relying on a simple godoc.

s-urbaniak commented:
@sgotti thanks, this looks much simpler than the initial version. Although it does not include the abstractions I was hoping for, it does LGTM, thanks!

lucab commented on common/distribution/docker.go:
```diff
@@ -0,0 +1,108 @@
+	}
+	fds2, err := FullDockerString(d2.ds)
+	if err != nil {
+		panic(err)
```
I think this panic (and the one above) shouldn't be here, and just return false instead.

lucab commented on common/distribution/aciarchive.go:
```diff
@@ -0,0 +1,100 @@
+// URI returns a copy of the Distribution URI
+func (a *ACIArchive) URI() *url.URL {
+	// Create a copy of the URL
+	u, err := url.Parse(a.u.String())
```
Why is this converting to string and then parsing it back again, just to create a copy? This is probably an expensive operation, and introduces an artificial panic point. I spotted this pattern in multiple point in this PR.

sgotti replied:
Currently it'll be possible to just copy by dereferencing but there's no guarantee that in a future url.URL won't change (adding a map or a slice) making them not deep copyable by derefernce. So just doing this seems the fastest way (I don't think it's so slow...) without relying on reflect.DeepCopy. The panic is to catch the error and if it fails there's really something wrong... Suggestions?

lucab replied:
I was actually thinking of deep-copying (I'm sorry I don't have performance numbers here, but I usually stay away from the re-assemble+re-parse pattern) but @s-urbaniak scared me with non-copyable things that may be in the opaque part. You two are the more knowledgeable in this area, so we can keep it as-is if no better options come to your minds.

lucab commented on common/distribution/helpers.go:
```diff
@@ -0,0 +1,136 @@
+
+// DockerString returns the docker string for a Docker distribution. It'll fail
+// if the distribution is not of type Docker.
+func DockerString(dist Distribution) (string, error) {
```
I'm not sure I understand why this function (and the surrounding ones) are here. They seem to be Docker-type specific and implement a runtime type-check. It looks like they would be attached as method to the corresponding type. What am I missing here?

sgotti replied:
Eh eh. That was the original implementation, I changed it trying to follow @s-urbaniak suggestion to remove type assertions (https://github.com/coreos/rkt/pull/3101#discussion_r76007576) but talking OOB he had a different idea that I don't completely get and I'm not sure it'll be feasible for the various requirements. So, for this implementation, I'd happily return to type assertions.

s-urbaniak commented:
Closing in favor of #3369
